### ISSUE TITLE
hasAnyRole gives wrong result

### ISSUE THREAD

**Issue by krgaurav1992** [spatie/laravel-permission#1223]
'$user->hasAnyRole(['admin', 'author'])` results in `true` even if user does not have that role.

Also, `$user->roles` gives all the array from the table `model_has_roles`.
`[
    {
        "id": 1,
        "name": "root_admin",
        "guard_name": "web",
        "created_at": "2019-08-26 10:58:59",
        "updated_at": "2019-08-26 10:58:59",
        "pivot": {
            "model_id": 1,
            "role_id": 1,
            "model_type": "App\\User"
        }
    },
    {
        "id": 2,
        "name": "admin",
        "guard_name": "web",
        "created_at": "2019-08-26 10:59:00",
        "updated_at": "2019-08-26 10:59:00",
        "pivot": {
            "model_id": 3,
            "role_id": 2,
            "model_type": "App\\User"
        }
    },
    {
        "id": 2,
        "name": "admin",
        "guard_name": "web",
        "created_at": "2019-08-26 10:59:00",
        "updated_at": "2019-08-26 10:59:00",
        "pivot": {
            "model_id": 6,
            "role_id": 2,
            "model_type": "App\\User"
        }
    },
    {
        "id": 3,
        "name": "author",
        "guard_name": "web",
        "created_at": "2019-08-26 10:59:00",
        "updated_at": "2019-08-26 10:59:00",
        "pivot": {
            "model_id": 3,
            "role_id": 3,
            "model_type": "App\\user"
        }
    },
    {
        "id": 3,
        "name": "author",
        "guard_name": "web",
        "created_at": "2019-08-26 10:59:00",
        "updated_at": "2019-08-26 10:59:00",
        "pivot": {
            "model_id": 5,
            "role_id": 3,
            "model_type": "App\\User"
        }
    }
]`

Anyone with similar issue ?

**Comment by krgaurav1992** [comment]
Figured out the issue. Apparently I was using the user object from `Auth::user()`. This was the reason `hasAnyRoles()` was sending incorrect result. I had to do `Auth::user()->fresh()` to fetch user as eloquent object and the result is correct now.

**Comment by krgaurav1992** [comment]
Reopening this issue because this is happpening with laravel 6.0 only and needs to be fixed.

**Comment by drbyte** [comment]
Please provide exact steps to recreate on a new Laravel app.

**Comment by scheMeZa** [comment]
I've noticed in Laravel 6; I have to fresh all my models when dealing with relationships. Does anyone know what the update is from Laravel 6 that are forcing us to use the fresh or unset method? And why it was changed?

**Comment by krgaurav1992** [comment]
The issue was with laravel and passport. Seems like they have fixed it now. I have ugraded laravel/framework and laravel/passort. It works correct now.

**Comment by krgaurav1992** [comment]
**It's here again !**

I have updated `laravel/framework` as well as `laravel/passport`. I am facing same issue again.

**Comment by drbyte** [comment]
Hi,

Help me recreate your problem. Assume I'm starting from a fresh Laravel app, by running the following:
```sh
laravel new permissions_test_app
cd permissions_test_app
git init
git add .
git commit -m "Fresh Laravel Install"

cp -n .env.example .env
sed -i '' 's/DB_CONNECTION=mysql/DB_CONNECTION=sqlite/' .env
sed -i '' 's/DB_DATABASE=laravel/#DB_DATABASE=laravel/' .env
touch database/database.sqlite

composer require spatie/laravel-permission
php artisan vendor:publish --provider="Spatie\Permission\PermissionServiceProvider"
git add .
git commit -m "Add Spatie Laravel Permissions package"
php artisan migrate:fresh

# Manually: add HasRoles trait to Users.php
```
Now, on top of this fresh new app, what exactly do I need to do to make your error happen in this app?
Obviously you'll need some migrations, some User code, some Controller code, etc. Please post all that is required, or a link to the repo where this simple new test app can be inspected.

**Comment by krgaurav1992** [comment]
@drbyte 
Please install `laravel/passport` by these commands:

`composer require laravel/passport`
`php artisan migrate`
`php artisan passport:install`

and then:
`
use Illuminate\Foundation\Auth\User as Authenticatable;
use Illuminate\Notifications\Notifiable;
use Laravel\Passport\HasApiTokens;

class User extends Authenticatable
{
    use HasApiTokens, Notifiable;
}
`

then, authenticate a user and generate a token using passport. for example:
`
public function authenticate (Request $request) {
        if (\Auth::attempt($request->only(['email', 'password']))) {
            $user = \App\User::where('email', $request->email)->first();
            return response()->json([
                'auth_token' => $user->createToken('user_token')->accessToken
            ]);
        }

        return response()->json(['message' => 'Invalid email or password'], 401);
    }
`

You must create 2 users and assign role to only one users just so to populate `model_has_roles` table. For example:
`
$role = \Spatie\Permission\Models\Role::findByName('root_admin', 'web');
$user->assignRole($role);
`

Now define a route with middleware using passport as: 
`
Route::middleware(['auth:api'])->group(function() {
     Route::get('users/me', 'UserController@me');
});
`

Now authenticate user who do not have any role and check for role using `hasRole` function. For example:
`
public function me() {
    $user = Auth::user();
    $user->hasRole(['admin']); // It always gives true even if user do not have any role.
    return response()->json($user);
}
`

**Comment by drbyte** [comment]
Are you intentionally using *both* `web` and `api` guards? Are they intentionally providing different permissions, or is it redundant to be using both in your app?
If you're really only needing `api`, then in your `config/auth.php` make `api` be your first (or only) guard.

**Comment by krgaurav1992** [comment]
I think I got the issue. You are right I am using `web` and `api` both the guards because I need both guards. I will try to do some work around for this. Thanks for your help.

**Comment by dennis-git** [comment]
you can pass guard name to $user->hasRole(['admin'], 'web') or $user->hasRole(['admin'], 'api') if you have v3.2 of the library

**Comment by dennis-git** [comment]
$user->hasAnyRole() doesn't check for guard_name but you check for it if you use $user->hasRole()

**Comment by RishabSwift** [comment]
I'm having this same issue. 

`auth()→user()→roles` gives `model_has_roles` roles, which is all of them rendering `auth()→user()→hasAnyRole([..])` useless.

@krgaurav1992 were you able to resolve this?

**Comment by drbyte** [comment]
Have you perchance defined a custom `roles` relationship that's overriding the one defined by the traits this package relies on?

**Comment by RishabSwift** [comment]
No, I have not defined a custom roles relationship nor overwritten any methods provided by the library.

I should mention though, this issue seems to be happening in a very specific page when I use it alongside Livewire. Like I have pagination with Livewire, and as soon as I paginate to the next page, all of the `if (user()→hasRole(...)` statements or `@can(..)` will return true if checking for roles. 

`(@role('...'))` will be true even though the logged in user doesn't have that role simply because *someone* on the `model_has_roles` table has that role.


Simply put, the `model_has_roles` table has 150 records for instance, and after pagination (or any Livewire action that re-renders the component), displaying `user()→roles` shows that the user has 150 roles (even though most are dupliacte)

**Comment by drbyte** [comment]
@RishabSwift Can you start another issue and set up a small demo app that replicates this problem, and link to it?
If it's specific to Livewire, then maybe we need to collaborate with them. 
If it's caused by something else in your app, then the demo app should help identify that and then we can find a solution.
